experimental/oak_async/src/executor.rs

```diff
@@ -0,0 +1,181 @@
+
+use core::{cell::RefCell, future::Future, task::Waker};
+use futures::{executor::LocalPool, task::LocalSpawnExt};
+use log::{debug, trace};
+use oak::{ChannelReadStatus, Handle, OakStatus, ReadHandle};
+use std::collections::HashMap;
+
+// thread local so `cargo test` can run tests in parallel.
+// Oak nodes are single-threaded.
+std::thread_local! {
+    static EXECUTOR: RefCell<Executor> = RefCell::new(Executor::default());
+}
+
+pub type ReaderId = u64;
+
+struct WaitingReader {
+    handle: Handle,
+    waker: Waker,
+}
+
+/// Global executor context.
+///
+/// [`with_executor`] provides a way to get a handle to the global executor instance.
+#[derive(Default)]
+pub struct Executor {
+    /// Used to assign a unique id to each reader. It is incremented by every newly created reader
+    next_reader_id: ReaderId,
+    /// Set of readers waiting for data. All handles in this map are polled in `wait_on_channels`.
+    // Why not HashMap<Handle, Vec<Waker>>? When a future asks to be removed from the waiting list
+    // (i.e. when it is dropped), we need to be able to remove their Waker from the map.
+    waiting_readers: HashMap<ReaderId, WaitingReader>,
+}
+
+impl Executor {
+    pub fn add_waiting_reader(&mut self, reader_id: ReaderId, handle: ReadHandle, waker: &Waker) {
+        trace!(
+            "Add waiting reader {} waiting on handle {}",
+            reader_id,
+            handle.handle
+        );
+        self.waiting_readers.insert(
+            reader_id,
+            WaitingReader {
+                handle: handle.handle,
+                waker: waker.clone(),
+            },
+        );
+    }
+
+    pub fn remove_waiting_reader(&mut self, reader_id: ReaderId) {
+        trace!("Remove waiting reader {}", reader_id);
+        if self.waiting_readers.remove(&reader_id).is_none() {
+            // This is usually not an error. If a Future is dropped as a result of it being woken up
+            // and then resolving, we expect the reader_id to not be present in the waiting set.
+            debug!(
+                "Attempted to remove waiting reader {}, but no such reader was in the waiting set",
+                reader_id
+            )
+        }
+    }
+
+    pub fn new_id(&mut self) -> ReaderId {
+        let id = self.next_reader_id;
+        // Collisions are technically possible, but only if one read is very slow and many async
+        // reads (2^64) are requested in quick succession.
+        self.next_reader_id = self.next_reader_id.wrapping_add(1);
+        id
+    }
+
+    pub fn none_waiting(&self) -> bool {
+        self.waiting_readers.is_empty()
+    }
+
+    /// Returns all handles for which reads are pending, and the id of one of the readers for that
+    /// channel.
+    ///
+    /// The `Vec`s returned are guaranteed to have the same length, and a handle at index `i`
+    /// corresponds to the reader id at index `i`.
+    pub fn pending_handles(&self) -> (Vec<ReadHandle>, Vec<ReaderId>) {
```

Comment:
This seems a bit awkward, is there no way of returning a Vec of some struct that keeps each ReadHandle next to its ReaderId? I think even a `Vec<(ReadHandle, ReaderId)>` would read better IMO.
In particular, unzipping here just to zip again in the caller seems less than ideal, plus the extra documentation explaining that those two Vecs are in fact same length and matching elements.

Reply:
Actually, that is close to what the original signature of this function was. I changed it based on the comment here: https://github.com/project-oak/oak/pull/1466#discussion_r487772512.

To clarify, we are not zipping these two `Vec`s together at the caller. Instead we pass the first vec to `wait_on_channels`, and then zip the result of that to the second vec.

Maybe I misunderstood one of the two comments, is there a signature that we can all agree on? To be honest both options seem fine to me.

Reply:
I think my previous comment was based on the code doing two iterations over the same collection, which seemed unnecessary when the `pending_handles()` method only had a single user.  Maybe mention in the function comment that the odd return type is there for convenience of feeding to `wait_on_channels` without further munging?